poetry/console/commands/plugin/add.py

```diff
@@ -0,0 +1,263 @@
+
+  - A single name (<b>requests</b>)
+  - A name and a constraint (<b>requests@^2.23.0</b>)
+  - A git url (<b>git+https://github.com/python-poetry/poetry.git</b>)
+  - A git url with a revision (<b>git+https://github.com/python-poetry/poetry.git#develop</b>)
+  - A git SSH url (<b>git+ssh://github.com/python-poetry/poetry.git</b>)
+  - A git SSH url with a revision (<b>git+ssh://github.com/python-poetry/poetry.git#develop</b>)
+  - A file path (<b>../my-package/my-package.whl</b>)
+  - A directory (<b>../my-package/</b>)
+  - A url (<b>https://example.com/packages/my-package-0.1.0.tar.gz</b>)\
+"""
+
+    def handle(self) -> int:
+        from pathlib import Path
+
+        import tomlkit
+
+        from cleo.io.inputs.string_input import StringInput
+        from cleo.io.io import IO
+
+        from poetry.core.semver.helpers import parse_constraint
+        from poetry.factory import Factory
+        from poetry.packages.project_package import ProjectPackage
+        from poetry.puzzle.provider import Provider
+        from poetry.repositories.installed_repository import InstalledRepository
+        from poetry.repositories.repository import Repository
+        from poetry.utils.env import EnvManager
+
+        plugins = self.argument("plugins")
+
+        # Plugins should be installed in the system env to be globally available
+        system_env = EnvManager.get_system_env()
+
+        env_dir = Path(
+            os.getenv("POETRY_HOME") if os.getenv("POETRY_HOME") else system_env.path
+        )
+
+        # We check for the plugins existence first.
+        if env_dir.joinpath("pyproject.toml").exists():
+            pyproject = tomlkit.loads(
+                env_dir.joinpath("pyproject.toml").read_text(encoding="utf-8")
+            )
+            poetry_content = pyproject["tool"]["poetry"]
+            existing_packages = self.get_existing_packages_from_input(
+                plugins, poetry_content, "dependencies"
+            )
+
+            if existing_packages:
+                self.notify_about_existing_packages(existing_packages)
+
+            plugins = [plugin for plugin in plugins if plugin not in existing_packages]
+
+        if not plugins:
+            return 0
+
+        plugins = self._determine_requirements(plugins)
+
+        # We retrieve the packages installed in the system environment.
+        # We assume that this environment will be a self contained virtual environment
+        # built by the official installer or by pipx.
+        # If not, it might lead to side effects since other installed packages
+        # might not be required by Poetry but still taken into account when resolving dependencies.
+        installed_repository = InstalledRepository.load(
+            system_env, with_dependencies=True
+        )
+        repository = Repository()
+
+        root_package = None
+        for package in installed_repository.packages:
+            if package.name in Provider.UNSAFE_PACKAGES:
```

Comment:
Note that we no longer consider pip or setuptools unsafe.

Reply:
Removed :-)